core/store/src/opener.rs

```diff
@@ -39,95 +131,216 @@ impl<'a> StoreOpener<'a> {
+            "{} RocksDB database",
+            if exists { "Opening an existing" } else { "Creating a new" }
+        );
+        let storage = self.open_storage(self.mode, exists.then_some(DB_VERSION))?;
+        if !exists {
             // Initialise newly created database by setting it’s version.
-            let store = crate::Store { storage };
-            version::set_store_version(&store, version::DB_VERSION)?;
-            Ok(crate::NodeStorage { storage: store.storage })
-        } else {
-            Ok(crate::NodeStorage { storage })
+            crate::version::set_store_version(&storage.get_store(Temperature::Hot), DB_VERSION)?;
+        }
+        Ok(storage)
+    }
+
+    /// Applies database migrations to the database.
+    fn apply_migrations(&self, db_version: DbVersion) -> Result<(), StoreOpenerError> {
+        if db_version == DB_VERSION {
+            return Ok(());
+        }
+        if self.mode == Mode::ReadOnly {
+            return Err(StoreOpenerError::DbVersionMismatchOnRead {
+                got: db_version,
+                want: DB_VERSION,
+            });
+        }
+        if db_version > DB_VERSION {
+            return Err(StoreOpenerError::DbVersionTooNew { got: db_version, want: DB_VERSION });
+        }
+
+        let migrator = match self.migrator {
+            Some(migrator) => {
+                if let Err(release) = migrator.check_support(db_version) {
+                    return Err(StoreOpenerError::DbVersionTooOld {
+                        got: db_version,
+                        want: DB_VERSION,
+                        latest_release: release,
+                    });
+                }
+                migrator
+            }
+            None => {
+                return Err(StoreOpenerError::DbVersionMismatch {
+                    got: db_version,
+                    want: DB_VERSION,
+                })
+            }
+        };
+
+        let snapshot = Snapshot::new(&self.db.path, &self.db.config)?;
+
+        for version in db_version..DB_VERSION {
+            tracing::info!(target: "near", path=%self.path().display(),
+                           "Migrating the database from version {version} to {}",
+                           version + 1);
+            let storage = self.open_storage(Mode::ReadWriteExisting, Some(version))?;
+            migrator.migrate(&storage, version).map_err(StoreOpenerError::MigrationError)?;
+            crate::version::set_store_version(&storage.get_store(Temperature::Hot), version + 1)?;
+        }
+
+        if cfg!(feature = "nightly") || cfg!(feature = "nightly_protocol") {
+            // Set some dummy value to avoid conflict with other migrations from
+            // nightly features.
+            let storage = self.open_storage(Mode::ReadWriteExisting, Some(DB_VERSION))?;
+            crate::version::set_store_version(&storage.get_store(Temperature::Hot), 10000)?;
         }
+
+        snapshot.remove();
```

Comment:
It’s because Snapshot::remove returns ().  If there’s an error cleaning up snapshot it logs an error but doesn’t propagate it.  Though now that I think about it, maybe it should return the error and opening should fail.